Use Date for last-modified attribute.

aptls opened the pull request:
Change the last-modified attribute on record elements to
marshal/unmarshal as a Date object. The conversion is done using a
custom XmlAdapter, so a ValidationEventHandler has been added when
unmarshalling using the DTD to report parse errors in the adapter.

Commit:
Use Date for last-modified attribute.

Change the last-modified attribute on record elements to
marshal/unmarshal as a Date object. The conversion is done using a
custom XmlAdapter, so a ValidationEventHandler has been added when
unmarshalling using the DTD to report parse errors in the adapter.

wiarlawd commented on src/com/google/enterprise/gsafeed/DateAdapter.java:
```diff
@@ -0,0 +1,59 @@
+  // Definition copied from GsaFeedFileMaker in adaptor library;
+  // format string extracted for access by other code if desired.
+  private static ThreadLocal<DateFormat> rfc822Format
+    = new ThreadLocal<DateFormat>() {
```
Indent +2, or move the = to the previous line.

wiarlawd commented on src/com/google/enterprise/gsafeed/FeedHelper.java:
```diff
@@ -82,6 +85,33 @@ public void fatalError(SAXParseException exception)
+              break;
+            default:
+              throw new IllegalArgumentException(
+                  "Unknown severity " + event.getSeverity());
```
Maybe add `+ " for " + event.getMessage()` for clarity?

wiarlawd commented on test/com/google/enterprise/gsafeed/RecordTest.java:
```diff
@@ -36,7 +43,7 @@ public void testRecord() throws Exception {
         + " displayurl='http://example.com' action='add'"
         + " mimetype='text/plain'"
-        + " last-modified='Tue, 6 Nov 2007 12:45:26 GMT'"
+        + " last-modified='Tue, 06 Nov 2007 12:45:26 +0000'"
```
We might need to revisit this, because I seem to remember using GMT, and mangling the formatter output in connector manager. Yeah, see Value.java in the connector/spi package. RFC 822 calls for "GMT" rather than "+0000" as produced by Java.

aptls replied:
Was there a reason not to use "z" instead?

wiarlawd commented on src/com/google/enterprise/gsafeed/DateAdapter.java:
```diff
@@ -0,0 +1,59 @@
+ * Converts between XML string representation of dates in RFC 822
+ * format and java.util.Date objects during marshal/unmarshal.
+ */
+public class DateAdapter extends XmlAdapter<String, Date> {
```
Does this need to be public?

wiarlawd commented on src/com/google/enterprise/gsafeed/DateAdapter.java:
```diff
@@ -0,0 +1,59 @@
+
+  // Definition copied from GsaFeedFileMaker in adaptor library;
+  // format string extracted for access by other code if desired.
+  private static ThreadLocal<DateFormat> rfc822Format
```
final

wiarlawd commented on src/com/google/enterprise/gsafeed/GsafeedHelper.java:
```diff
@@ -40,6 +41,10 @@ public void setErrorHandler(ErrorHandler errorHandler) {
     this.errorHandler = errorHandler;
   }
 
+  public void setValidationEventHandler(ValidationEventHandler eventHandler) {
```
Is this really part of the API? What's the use case for this (beyond the tests)?

aptls replied:
Same as the reason for exposing the error handler: I don't want to assume that the simplest possible implementation is all that's ever called for. Someone could want a different sort of logging or something, or have a use case where they don't care if the dates are wrong. It's also possible to switch to use javax.xml.bind.util.ValidationEventCollector to accumulate events instead of failing immediately.

wiarlawd reviewed (commented):
Minor stuff, except for "GMT".

aptls commented:
How do you want to handle messages from the validation handler (and others in future), with print statements, or logging, or something else?

wiarlawd commented on src/com/google/enterprise/gsafeed/FeedHelper.java:
```diff
@@ -82,6 +85,33 @@ public void fatalError(SAXParseException exception)
+              throw new IllegalArgumentException(
+                  "Unknown severity " + event.getSeverity());
+          }
+          // TODO(aptls): find out if we want logging, System.out, ...
```
A TODO seemed OK, but I had a lean toward logging at INFO or higher, which is written to the console by default, but configurable, so maybe the best of both worlds.

Commit:
Code review changes.

Use "GMT" instead of "+0000" in dates.

Use java.util.logging for validation messages. Add a JUnit rule to
disable console logging during tests to keep test output readable.

wiarlawd commented on test/com/google/enterprise/gsafeed/RecordTest.java:
```diff
@@ -519,6 +524,24 @@ public void testLastModifiedDifferentFormatWithValidation()
     new GsafeedHelper().unmarshalWithDtd(feed);
   }
 
+  // Add an even more straightforward date format test here.
```
Is this a TODO(aptls)?

aptls replied:
No, just a poorly-phrased comment. I wanted to include a basic check of the formatting with no extra layers.

wiarlawd commented on src/com/google/enterprise/gsafeed/DateAdapter.java:
```diff
@@ -0,0 +1,59 @@
+ * format and java.util.Date objects during marshal/unmarshal.
+ */
+class DateAdapter extends XmlAdapter<String, Date> {
+  public static final String FORMAT = "EEE, dd MMM yyyy HH:mm:ss z";
```
Maybe s/dd/d/ since this changed some test data from "6 Nov 2007" to "06 Nov 2007"? Is the single-digit day from GSA feed files?

aptls replied:
The single-digit day is from the feed example in the feed developer's guide, which is the sample data I started from for the first tests. The two-digit form is in the feed I downloaded from a GSA. The two-digit form is also used in GsaFeedFileMaker.

wiarlawd reviewed (approved):
LGTM++ with nits

PapaJoltOmega reviewed (approved):
Thank you

Commit:
Make comment clearer.
